pyogrio/tests/test_core.py

```diff
@@ -611,3 +612,94 @@ def test_error_handling_warning(capfd, naturalearth_lowres):
         read_info(naturalearth_lowres, INVALID="YES")
 
     assert capfd.readouterr().err == ""
+
+
+def test_vsimem_listtree(naturalearth_lowres):
+    # Prepare test data in /vsimem
+    meta, _, geometry, field_data = read(naturalearth_lowres)
+    meta["spatial_index"] = False
+    meta["geometry_type"] = "MultiPolygon"
+    test_file_path = Path("/vsimem/pyogrio_test_naturalearth_lowres.gpkg")
+    test_dir_path = Path(f"/vsimem/pyogrio_dir_test/{naturalearth_lowres.stem}.gpkg")
+
+    write(test_file_path, geometry, field_data, **meta)
+    write(test_dir_path, geometry, field_data, **meta)
+
+    # Check basic listtree
+    files = vsi_listtree("/vsimem/")
+    assert test_file_path.as_posix() in files
+    assert test_dir_path.as_posix() in files
+
+    # Check listtree with pattern
+    files = vsi_listtree("/vsimem/", pattern="pyogrio_dir_test*.gpkg")
+    assert test_file_path.as_posix() not in files
+    assert test_dir_path.as_posix() in files
+
+    files = vsi_listtree("/vsimem/", pattern="pyogrio_test*.gpkg")
+    assert test_file_path.as_posix() in files
+    assert test_dir_path.as_posix() not in files
+
+    # Clean up
+    vsi_unlink(test_file_path)
+    vsi_rmtree(test_dir_path.parent)
+
+
+@pytest.mark.parametrize(
+    "path, is_fixture, exception, error_match",
+    [
+        (
+            "naturalearth_lowres_vsimem",
+            True,
+            NotADirectoryError,
+            "Path is not a directory",
+        ),
+        ("/vsimem/non-existent", False, FileNotFoundError, "Path does not exist"),
+        ("/vsimem", False, OSError, "path to in-memory file or directory is required"),
+        ("/vsimem/", False, OSError, "path to in-memory file or directory is required"),
+    ],
+)
+def test_vsimem_rmtree_error(path, is_fixture, exception, error_match, request):
+    if is_fixture:
+        path = request.getfixturevalue(path)
+
+    with pytest.raises(exception, match=error_match):
+        vsi_rmtree(path)
+
+    # If path was a fixture, the temp file should still exists.
+    if is_fixture:
+        assert path.as_posix() in vsi_listtree("/vsimem")
+
+
+def test_vsimem_rmtree_unlink(naturalearth_lowres):
```

Comment:
This one seems for a large part the same as `test_vsimem_listtree` above? Maybe those could be combined (because to test listtree, you also need to rmtree/unlink the result, so then maybe could just test them together)

Reply:
They are indeed very similar... I'll merge them